File: src/blenderio.js
const child_process = require("child_process");
const net = require("net");

class IO_IN {
    constructor(initializer) {
        this.status = initializer.status;
        this.data = initializer.data;
    }
}
exports.IO_IN = IO_IN;

class IO_OUT {
    constructor(status, data = null) {
        this.status = status;
        this.data = data;
    }
    json() {
        return JSON.stringify({ status: this.status, data: this.data }) + "\n";
    }
}
exports.IO_OUT = IO_OUT;

class BlenderIO {
    constructor(
        script = `./src/py/core.py`,
        blender_executable = `${process.cwd()}\\blender\\blender.exe`,
        pref = {
            js_log_in: true,
            js_log_out: true,
            python_log_in: true,
            python_log_out: true,
            blender_background: true,
            keep_blender_open: false,
            render_engine: "EEVEE",
            render_samples: 32,
        }
    ) {
        this.pref = pref;
        this.STREAM_FIFO = [];
        this.SOCKET_FIFO = [];
        this.SOCKET_BUFF = "";
        this.socket_io = null;
        this.MODE = "STREAM";
        this.IPv4 = "127.0.0.1";
        this.PORT = null;

        let args = ["-P", script];
        if (this.blender_background) {
            args.unshift("-b");
        }
        this.blender_process = child_process.spawn(blender_executable, args, {
            cwd: process.cwd(),
        });
        this.stdout = this.blender_process.stdout;
        this.stdout.on("data", raw_data => {
            for (const sub_block of raw_data.toString().trim().split("\n")) {
                try {
                    this.STREAM_FIFO.push(JSON.parse(sub_block));
                } catch (e) {}
            }
        });
        this.stdin = this.blender_process.stdin;
    }
    kill() {
        if (!this.keep_blender_open) {
            this.blender_process.kill();
            this.socket_io.destroy();
        } else {
            this.call(new IO_OUT("Detach"));
        }
    }
    async begin() {
        let mess = await this.read_status("READY");
        this.MODE = "SOCKET";
        this.PORT = mess.data.port;
        this.socket_io = new net.Socket();
        await new Promise(resolve => {
            if (!this.socket_io.connecting) {
                this.socket_io.connect(this.PORT, this.IPv4, () => {
                    resolve(true);
                });
            }
        });
        this.socket_io.on("data", raw_data => {
            for (const sub_block of raw_data.toString().trim().split("\n")) {
                try {
                    this.SOCKET_FIFO.push(JSON.parse(sub_block));
                } catch (e) {}
            }
        });
        this.write(
            new IO_OUT("READY SOCKET", {
                python_log_in: this.pref.python_log_in,
                python_log_out: this.pref.python_log_out,
                render_engine: this.pref.render_engine,
                render_samples: this.pref.render_samples,
                render_dpi: this.pref.render_dpi,
            })
        );
        mess = await this.read_status("WAITING SOCKET");
    }
    _read_source(fifo) {
        function pull(resolve) {
            if (fifo.length > 0) {
                resolve(new IO_IN(fifo.shift()));
            } else {
                setTimeout(() => pull(resolve), 10);
            }
        }
        return new Promise(resolve => {
            pull(resolve);
        });
    }
    async read() {
        let mess;
        if (this.MODE == "STREAM") {
            mess = await this._read_source(this.STREAM_FIFO);
        } else if (this.MODE == "SOCKET") {
            mess = await this._read_source(this.SOCKET_FIFO);
        }
        if (this.pref.js_log_io) {
            console.log(mess);
        }
        return mess;
    }
    async read_status(status) {
        let mess = await this.read();
        if (mess.status != status) {
            throw Error(`BlenderIO connection error ${mess.data.trace}`);
        }
        return mess;
    }
    _write_stream(io_out_message) {
        this.stdin.write(io_out_message.json());
    }
    _write_socket(io_out_message) {
        this.socket_io.write(io_out_message.json());
    }
    write(io_out_message) {
        if (this.pref.js_log_out) {
            console.log(io_out_message);
        }
        if (this.MODE == "STREAM") {
            this._write_stream(io_out_message);
        } else if (this.MODE == "SOCKET") {
            this._write_socket(io_out_message);
        }
    }
    async call(io_out_call, status_ok = "OK") {
        this.write(io_out_call);
        let mess = await this.read_status(status_ok);
        return mess;
    }
}
exports.BlenderIO = BlenderIO;

File: src/py/blenderio.py
import socket
import json
import datetime
import sys
from typing import TextIO
from src.py import Singleton


class IO_OBJECT:

    status: str
    data: dict

    def json(self):
        return (
            json.dumps(
                {
                    "status": self.status,
                    "data": self.data,
                    "time": datetime.datetime.now().strftime("%Y.%m.%d.%H.%M.%S.%f"),
                }
            )
            + "\n"
        )

    def __str__(self) -> str:
        return (
            self.__class__.__name__
            + " "
            + json.dumps(
                {
                    "status": self.status,
                    "data": self.data,
                    "time": datetime.datetime.now().strftime("%Y.%m.%d-%H.%M.%S.%f"),
                },
                indent="  ",
            )
        )


class IO_OUT(IO_OBJECT):
    def __init__(self, status: str, data: dict = None) -> None:
        self.status = status
        self.data = data


class IO_IN(IO_OBJECT):
    status: str
    data: dict

    def __init__(self, initializer: str) -> None:
        self.status = initializer["status"]
        self.data = initializer["data"]


class BlenderIO(Singleton):

    PORT: int = 3568
    IPv4: str = "127.0.0.1"
    MODE: str = "STREAM"

    def __init__(self) -> None:
        self.stdout: TextIO = sys.stdout
        self.stdin: TextIO = sys.stdin
        self.socket: socket.socket = None
        self.socket_io: socket.socket = None
        self.python_log_in: bool = True
        self.python_log_out: bool = True
        self.log_file = open(
            f"./temp/blenderio-{datetime.datetime.now().strftime('%Y.%m.%d-%H-%M-%S-%f')}.log",
            "a",
        )

    def __del__(self) -> None:
        if self.socket is not None:
            self.socket.close()
        if self.socket_io is not None:
            self.socket_io.close()

    def log(self, *args) -> None:
        self.log_file.write(" ".join(str(a) for a in args) + "\n")
        self.log_file.flush()

    def begin(self) -> None:
        self.socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        while True:
            try:
                self.socket.bind((self.IPv4, self.PORT))
                break
            except Exception:
                self.PORT += 1
                continue
        self.socket.listen()
        self.write(IO_OUT("READY", {"port": self.PORT}))
        self.MODE = "SOCKET"
        self.socket_io, address = self.socket.accept()
        mess = self.read()
        if mess.status != "READY SOCKET":
            raise Exception()
        self.python_log_in = mess.data["python_log_in"]
        self.python_log_out = mess.data["python_log_out"]
        self.render_dpi = mess.data["render_dpi"]
        self.render_engine = mess.data["render_engine"]
        self.render_samples = mess.data["render_samples"]
        self.write(IO_OUT("WAITING SOCKET"))

    def _write_stream(self, message: IO_OUT) -> None:
        self.stdout.flush()
        self.stdout.write(message.json())
        self.stdout.flush()

    def _write_socket(self, message: IO_OUT) -> None:
        self.socket_io.sendall(message.json().encode("utf-8"))

    def write(self, message: IO_OUT) -> None:
        if self.python_log_out:
            self.log(message)
        if self.MODE == "STREAM":
            self._write_stream(message)
        elif self.MODE == "SOCKET":
            self._write_socket(message)

    def _read_steam(self) -> IO_IN:
        block = self.stdin.readline().strip()
        while not block:
            block = self.stdin.readline().strip()
        return IO_IN(json.load(block))

    def _read_socket(self) -> IO_IN:
        bytebuffer = bytearray()
        while True:
            letter = self.socket_io.recv(1)
            if letter == b"\n":
                break
            else:
                bytebuffer += letter
        bytebuffer = bytebuffer.decode("utf-8")
        return IO_IN(json.loads(bytebuffer))

    def read(self) -> IO_IN:
        if self.MODE == "STREAM":
            message = self._read_steam()
        elif self.MODE == "SOCKET":
            message = self._read_socket()
        if self.python_log_in:
            self.log(message)
        return message
